### PR TITLE
support TSL configuration for HAproxy

### PR DESCRIPTION
**Pull request by adk3798**
This should roughly provide what we need. Here is an example, showing the spec file provided and the resulting haproxy.cfg: https://pastebin.com/7vbSuUQx

However, I still need some more information to finish this. 

First, in terms of verification, it seems pretty clear that if the user wants to use TSL (which is indicated by them including the ha_proxy_frontend_ssl_certificate in the spec) that they NEED to have the ssl frontend certificate and the ssl frontend port, but I'm not sure if we are supposed to also require the ssl dh param, at least one ssl cipher or at least one ssl option. Right now, I have it requiring something be put for all TSL related fields if the certificate is provided but I'm not sure that's right.

Second, in the example template provided https://raw.githubusercontent.com/ceph/ceph-ansible/master/roles/ceph-rgw-loadbalancer/templates/haproxy.cfg.j2 in the bind  lines of the "frontend rgw-frontend" section the ip provided is just * where as in our templates so far we use a user provided virtual ip address. Should we still use the virtual ip address in the case of TSL or swap to using * in that case?

Lastly, some reasonable sample values for these fields would be helpful for testing. As you can see from my example in the pastebin, I don't have any background knowledge related to this feature and just put sort of random values down.

Signed-off-by: Adam King <adking@redhat.com>

### CODE REVIEW

**Comment by jmolmo** [comment]
> First, in terms of verification, it seems pretty clear that if the user wants to use TSL (which is indicated by them including the ha_proxy_frontend_ssl_certificate in the spec) that they NEED to have the ssl frontend certificate and the ssl frontend port, but I'm not sure if we are supposed to also require the ssl dh param, at least one ssl cipher or at least one ssl option. Right now, I have it requiring something be put for all TSL related fields if the certificate is provided but I'm not sure that's right.

Use the values pointed in:
https://github.com/ceph/ceph-ansible/blob/master/roles/ceph-rgw-loadbalancer/defaults/main.yml





You can find more information about possibilities here:
https://cbonte.github.io/haproxy-dconv/2.0/configuration.html#ssl-default-bind-ciphers

The option and ciphers lines can have a big amount of different things ... for example:

`ssl-default-bind-ciphers ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES256-GCM-SHA384:DHE-RSA-AES128-GCM-SHA256:DHE-DSS-AES128-GCM-SHA256:kEDH+AESGCM:ECDHE-RSA-AES128-SHA256:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA:ECDHE-ECDSA-AES128-SHA:ECDHE-RSA-AES256-SHA384:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA:ECDHE-ECDSA-AES256-SHA:DHE-RSA-AES128-SHA256:DHE-RSA-AES128-SHA:DHE-DSS-AES128-SHA256:DHE-RSA-AES256-SHA256:DHE-DSS-AES256-SHA:DHE-RSA-AES256-SHA:!aNULL:!eNULL:!EXPORT:!DES:!RC4:!3DES:!MD5:!PSK`

`ssl-default-bind-options ssl-min-ver TLSv1.0 no-tls-tickets`

Note: pay attention that _ciphers_ are separated using ":" , and _options_ are separated by space.

For the moment we are not  to verify that the cipher/options  are correct. Just get the values and render them properly in the template. (If haproxy starts with TLS protocol enabled then they are ok :-)

**Comment by jmolmo** [comment]
> Should we still use the virtual ip address in the case of TSL or swap to using * in that case?

use the virtual IP

File: src/python-common/ceph/deployment/service_spec.py [hunk 1]
```diff
@@ -849,6 +859,19 @@ def validate(self):
...
             raise ServiceSpecValidationError(
                 'Cannot add HA_RGW: No Keepalived Password specified')
+        if self.ha_proxy_frontend_ssl_certificate:
+            if not self.ha_proxy_frontend_ssl_port:
+                raise ServiceSpecValidationError(
+                    'Cannot add HA_RGW: Specified Ha Proxy Frontend SSL Certificate but no SSL Port')
+            if not self.ha_proxy_ssl_dh_param:
+                raise ServiceSpecValidationError(
+                    'Cannot add HA_RGW: Specified Ha Proxy Frontend SSL Certificate but not SSL dh param')
+            if not self.ha_proxy_ssl_ciphers:
```
Comment (jmolmo) [inline src/python-common/ceph/deployment/service_spec.py:869]: Add checks in order to verify that values provided are "inside" the ones we support

**Comment by adk3798** [comment]
Changed this so now, if the ssl certificate is supplied, it still requires the ssl frontend port but does not require the dh param, ssl ciphers or ssl options be provided. Those 3 parameters will be added to the config only if the ssl certificate is present and they are individually set, otherwise they are left out.

**Review by jmolmo — APPROVED** [review]

**Pull request merged**